### PR TITLE
Deprecating RestSharp in Download Clients

### PR DESCRIPTION
**Pull request by Taloth**
@kayone @markus101 

Lots of stuff happening here. Isolated several big chunks into separate commits.

I'm generally happy with how most of the stuff is now in HttpRequestBuilder. And curl post works, so that's nice too.

AddFormParameter & AddFormUpload, converted to a neat multipart formdata.

I also removed Uri and replaced it with a custom HttpUri, at one point I started to wonder if that was smart. There might be odd urls I handle wrong. But I was kinda pissed off at Uri for it's weird behavior. At first I just wanted to encapsulate Uri, so I could prevent ppl from using the wrong methods (ToString() & AbsoluteUri), but eventually opted to roll my own.

Lemme know what you think and feel free to shoot at it.

Proxies using RestSharp:

- [x] Sabnzbd
- [x] NzbGet
- [x] Deluge
- [x] QbitTorrent
- [x] NzbVortexProxy
- [x] TransmissionProxy
- [x] UtorrentProxy

There are also several notifications using RestSharp, not sure if we should update those too. And finally the integration tests are using RestSharp.

### CODE REVIEW

File: src/NzbDrone.Common.Test/Http/HttpClientFixture.cs [hunk 1]
```diff
@@ -58,18 +58,18 @@ public void should_execute_typed_get()
...
-            response.Resource.Url.Should().Be(request.Url.ToString());
+            response.Resource.Url.Should().Be(request.Url.FullUri);
         }
 
         [Test]
         public void should_execute_simple_post()
         {
             var request = new HttpRequest("http://eu.httpbin.org/post");
-            request.Body = "{ my: 1 }";
+            request.SetContent("{ my: 1 }");
```
Comment (markus101) [inline src/NzbDrone.Common.Test/Http/HttpClientFixture.cs]: Store in variable so we don't need to build it twice?

File: src/NzbDrone.Common.Test/Http/HttpClientFixture.cs [hunk 2]
```diff
@@ -322,10 +322,10 @@ public void should_parse_malformed_cloudflare_cookie(string culture)
             {
                 // the date is bad in the below - should be 13-Jul-2016
                 string malformedCookie = @"__cfduid=d29e686a9d65800021c66faca0a29b4261436890790; expires=Wed, 13-Jul-16 16:19:50 GMT; path=/; HttpOnly";
-                string url = "http://eu.httpbin.org/response-headers?Set-Cookie=" +
-                    System.Uri.EscapeUriString(malformedCookie);
+                var requestSet = new HttpRequestBuilder("http://eu.httpbin.org/response-headers")
+                    .AddQueryParam("Set-Cookie", malformedCookie).
```
Comment (markus101) [inline src/NzbDrone.Common.Test/Http/HttpClientFixture.cs]: Trailing `.`

File: src/NzbDrone.Common/Http/HttpClient.cs [hunk 3]
```diff
@@ -133,7 +134,7 @@ private CookieContainer PrepareRequestCookies(HttpRequest request)
                 return cookieContainer;
             }
         }
-
+
```
Comment (markus101) [inline src/NzbDrone.Common/Http/HttpClient.cs]: Accident I'm sure :smile:

File: src/NzbDrone.Core/Download/Clients/Deluge/DelugeProxy.cs [hunk 4]
```diff
@@ -157,134 +159,134 @@ public void SetLabel(string hash, string label, DelugeSettings settings)
...
-                }
+                url = string.Format(@"{0}://{1}:{2}/json", protocol, settings.Host, settings.Port);
             }
 
+            var builder = new JsonRpcRequestBuilder(url);
+
+            var httpRequest = new HttpRequest(url, HttpAccept.Json);
+
+#warning FIXME: Had a 15sec timeout.
+            //restClient.Timeout = 15000;
```
Comment (markus101) [inline src/NzbDrone.Core/Download/Clients/Deluge/DelugeProxy.cs]: Problem solved!

File: src/NzbDrone.Core/Download/Clients/Deluge/DelugeProxy.cs [hunk 5]
```diff
@@ -157,134 +159,134 @@ public void SetLabel(string hash, string label, DelugeSettings settings)
...
-            if (arguments != null)
+                return Json.Deserialize<JsonRpcResponse<TResult>>(response.Content);
+            }
+            catch (HttpException ex)
             {
-                data.Add("params", arguments);
+                if (ex.Response.StatusCode == HttpStatusCode.RequestTimeout)
+                {
+                    _logger.Debug("Deluge timeout during request, daemon connection may have been broken. Attempting to reconnect.");
+                    return new JsonRpcResponse<TResult>()
```
Comment (markus101) [inline src/NzbDrone.Core/Download/Clients/Deluge/DelugeProxy.cs]: Should we throw an exception here then reconnect?

File: src/NzbDrone.Core/Download/Clients/Nzbget/NzbgetProxy.cs [hunk 6]
```diff
@@ -160,68 +147,46 @@ public void RetryDownload(string id, NzbgetSettings settings)
...
             {
-                throw new DownloadClientException("Unable to connect to NzbGet. " + response.ErrorException.Message, response.ErrorException);
+                response = _httpClient.Execute(httpRequest);
             }
-
-            if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
+            catch (HttpException ex)
             {
-                throw new DownloadClientException("Authentication failed for NzbGet, please check your settings", response.ErrorException);
+                if (ex.Response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
```
Comment (markus101) [inline src/NzbDrone.Core/Download/Clients/Nzbget/NzbgetProxy.cs]: Import System.Net

File: src/NzbDrone.Core/Download/Clients/Sabnzbd/SabnzbdProxy.cs [hunk 7]
```diff
@@ -85,45 +76,48 @@ public string GetVersion(SabnzbdSettings settings)
...
+                request.AddQueryParam("category", category);
             }
 
-            var response = ProcessRequest(request, action, settings);
+            var response = ProcessRequest(request, settings);
+
             return Json.Deserialize<SabnzbdHistory>(JObject.Parse(response).SelectToken("history").ToString());
         }
 
         public string RetryDownload(string id, SabnzbdSettings settings)
```
Comment (markus101) [inline src/NzbDrone.Core/Download/Clients/Sabnzbd/SabnzbdProxy.cs]: Do we use this any more? Could probably nuke it.

**Comment by markus101** [comment]
Haven't tested it out, but looks good.

File: src/NzbDrone.Core/Download/Clients/Deluge/DelugeProxy.cs (hunk 4 above)
Comment (Taloth) [inline src/NzbDrone.Core/Download/Clients/Deluge/DelugeProxy.cs]: Euh, no, coz HttpClient/Request doesn't hav a Timeout parameter (yet).

File: src/NzbDrone.Core/Download/Clients/Deluge/DelugeProxy.cs (hunk 5 above)
Comment (Taloth) [inline src/NzbDrone.Core/Download/Clients/Deluge/DelugeProxy.cs]: No, we fake a deluge response saying it need auth, then we deal with reauth later in the other ProcessResponse method.
(This particular scenario only triggers if the web api lost it's connection with the daemon rpc.)

(Yes, Deluge web api is aweful)

File: src/NzbDrone.Core/Download/Clients/Sabnzbd/SabnzbdProxy.cs (hunk 7 above)
Comment (Taloth) [inline src/NzbDrone.Core/Download/Clients/Sabnzbd/SabnzbdProxy.cs]: Not being used but tbh, not sure if we should nuke it.

File: src/NzbDrone.Common/Http/HttpHeader.cs [hunk 8]
```diff
@@ -4,37 +4,92 @@
 using System.Collections.Specialized;
 using System.Text;
 using NzbDrone.Common.Extensions;
+using System.Collections;
+using System.Globalization;
 
 namespace NzbDrone.Common.Http
 {
-    public class HttpHeader : Dictionary<string, object>
+    public class HttpHeader : NameValueCollection, IEnumerable<KeyValuePair<string, string>>, IEnumerable
```
Comment (kayone) [inline src/NzbDrone.Common/Http/HttpHeader.cs]: Why not a dictionary?

File: src/NzbDrone.Common/Http/HttpRequestBuilder.cs [hunk 9]
```diff
@@ -1,33 +1,123 @@
...
+using NzbDrone.Common.Extensions;
 
 namespace NzbDrone.Common.Http
 {
     public class HttpRequestBuilder
     {
-        public Uri BaseUri { get; private set; }
-        public bool SupressHttpError { get; set; }
+        public HttpMethod Method { get; set; }
+        public HttpAccept HttpAccept { get; set; }
```
Comment (kayone) [inline src/NzbDrone.Common/Http/HttpRequestBuilder.cs]: shouldn't this be part of `HttpHeader`?

File: src/NzbDrone.Common/Http/HttpHeader.cs (hunk 8 above)
Comment (Taloth) [inline src/NzbDrone.Common/Http/HttpHeader.cs]: In some cases multiple headers are allowed. That's why the webrequestheadercollection uses namevaluecollection as well.

**Pull request merged**